src/Estimators/InputSection.h

```diff
@@ -0,0 +1,116 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2021 QMCPACK developers.
+//
+// File developed by: Jaron T. Krogel, krogeljt@ornl.gov, Oak Ridge National Laboratory
+//
+//////////////////////////////////////////////////////////////////////////////////////
+#ifndef QMCPLUSPLUS_INPUTSECTION_H
+#define QMCPLUSPLUS_INPUTSECTION_H
+
+#include <typeinfo>
+#include <any>
+#include <unordered_set>
+#include <unordered_map>
+
+#include "Configuration.h"
+#include "OhmmsData/ParameterSet.h"
+#include "Containers/OhmmsPETE/TinyVector.h"
+
+#include "Message/UniformCommunicateError.h"
+
+
+namespace qmcplusplus
+{
+
+
+class InputSection
+{
+
+public:
+  using Real               = QMCTraits::RealType;
+  using PosType            = QMCTraits::PosType;
+  static constexpr int DIM = QMCTraits::DIM;
+
+protected:
+  // Internal data below comprise the input specification.
+  //   Most apply attributes to input variables.
+  //   Enables minimal listing of variable classification and default values in derived classes.
+  //   Expand later to include allowed_values for input correctness checking
+  std::string section_name;                    // name of the input section
+
+  std::unordered_set<std::string> attributes;  // list of attribute variables
+  std::unordered_set<std::string> parameters;  // list of parameter variables
+  std::unordered_set<std::string> required;    // list of required variables
+
+  std::unordered_set<std::string> strings;     // list of string variables
+  std::unordered_set<std::string> bools;       // list of boolean variables
+  std::unordered_set<std::string> integers;    // list of integer variables
+  std::unordered_set<std::string> reals;       // list of real variables
+  
+  std::unordered_map<std::string, std::any> default_values; // default values for optional variables
+  
+private:
+  // Storage for variable values read from XML, etc.
+  std::unordered_map<std::string, std::any> values;
+  
+public:
+  // Query functions
+  bool is_attribute(std::string name){return attributes.find(name)!=attributes.end();};
+  bool is_parameter(std::string name){return parameters.find(name)!=parameters.end();};
+  bool is_required(std::string name){return required.find(name)!=required.end();};
+
+  bool is_string(std::string name){return strings.find(name)!=strings.end();};
+  bool is_bool(std::string name){return bools.find(name)!=bools.end();};
+  bool is_integer(std::string name){return integers.find(name)!=integers.end();};
+  bool is_real(std::string name){return reals.find(name)!=reals.end();};
+
+  bool has_default(std::string name){return default_values.find(name)!=default_values.end();};
+  bool has(std::string name){return values.find(name)!=values.end();};
```

Comment:
please mark all the above is_* has_* functions const.

Reply:
Done